### PR TITLE
test: setup testing

### PR DESCRIPTION
**Pull request by greatertomi**

### CODE REVIEW

**Comment by greatertomi** [comment]
Why did you change all your responses from `.send` to `.json`?

File: src/test/factory.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  private async startup() {
+    try {
+      const connection = await mongoose.connect(MONGO_URL);
+      this._connection = connection.connection;
+      this._app = express();
+      this._app.use(express.json());
+      this._app.use(express.urlencoded({ extended: true }));
+      this._app.use('/', routes);
```
Comment (greatertomi) [inline src/test/factory.ts:38]: This is why the errors were not showing up properly in the test.

```suggestion
      this._app.use('/', routes);
      this._app.use(errorHandler);
```

File: src/test/url.test.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        .send({ url: 'https://www.mail.google.com/mail/u/0/#inbox/FMfcgzGxTNvvQTsMtWlbcjkNvlmPkjnH' });
+
+      expect(response.status).toBe(HTTP_ERRORS.CREATED);
+      expect(response.body.message).toBe('Url shortened successfully');
+      expect(response.body.data).toHaveProperty('shortUrl');
+    });
+
+    it('should return an error if the request body is invalid', async () => {
+      const res1 = await factory.app.post('/shorten').send({});
+      expect(res1.status).toBe(HTTP_ERRORS.BAD_REQUEST);
```
Comment (greatertomi) [inline src/test/url.test.ts:56]:
```suggestion
      expect(res1.status).toBe(HTTP_ERRORS.BAD_REQUEST);
      expect(res1.body.errors).toHaveLength(2);

      const res2 = await factory.app.post('/shorten').send({ url: 'google beats everything' });
      expect(res2.status).toBe(HTTP_ERRORS.BAD_REQUEST);
      expect(res2.body.errors).toHaveLength(1);
```

File: src/test/url.test.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+      expect(res2.status).toBe(HTTP_ERRORS.OK);
+      expect(res2.body.originalUrl).toBe('https://www.youtube.com/watchme?v=xZLKALpvdBE');
+      expect(res2.body).toHaveProperty('shortUrl');
+    });
+
+    it('should return an error if the url does not exist', async () => {
+      const nonExistentId = '609c6a0f0ac8cd1e58fd6d87';
+      const response = await factory.app.get(`/urls/${nonExistentId}`);
+      expect(response.status).toBe(HTTP_ERRORS.NOT_FOUND);
```
Comment (greatertomi) [inline src/test/url.test.ts:117]: Add more assertions to this error test

File: src/test/url.test.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      expect(res1.body.message).toBe('Url updated successfully');
+      expect(res1.body.data.originalUrl).toBe('https://gist.github.com/a3ae20d4747b7775f008071aa6587e90');
+    });
+
+    it('should return an error if the url id does not exist', async () => {
+      const nonExistentId = '609c6a0f0ac8cd1e58fd6d87';
+      const response = await factory.app
+        .put(`/urls/${nonExistentId}`)
+        .send({ url: 'https://www.boredbutton.com/lorem/ipsum/dolor/sit/amet' });
+      expect(response.status).toBe(HTTP_ERRORS.NOT_FOUND);
```
Comment (greatertomi) [inline src/test/url.test.ts:151]: Add more assertions to this error test

File: src/test/url.test.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    it('should return an error if the request body is invalid', async () => {
+      const res1 = await factory.app
+        .post('/shorten')
+        .send({ url: 'https://www.testwebsite.com/lorem/ipsum/dolor/sit/amet', customName: 'why not' });
+      expect(res1.status).toBe(HTTP_ERRORS.CREATED);
+
+      const urlId = res1.body.data.id;
+      const res2 = await factory.app.put(`/urls/${urlId}`).send({});
+      expect(res2.status).toBe(HTTP_ERRORS.BAD_REQUEST);
```
Comment (greatertomi) [inline src/test/url.test.ts:163]: Add more assertions to this error tests

File: src/test/url.test.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      const resp = await factory.app
+        .post('/shorten')
+        .send({ url: 'https://www.madeupwebsitefortesting.com', customName: 'fast times' });
+      expect(resp.status).toBe(HTTP_ERRORS.CREATED);
+
+      const urlId = resp.body.data.id;
+      const res2 = await factory.app
+        .put(`/urls/${urlId}`)
+        .send({ url: 'https://www.madeupwebsitefortestingtimestwo.com', customName: 'fast times' });
+      expect(res2.status).toBe(HTTP_ERRORS.BAD_REQUEST);
```
Comment (greatertomi) [inline src/test/url.test.ts:183]: Add more assertions to this error tests

File: src/test/url.test.ts [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+      const nonExistentId = '609c6a0f0ac8cd1e58fd6d87';
+      const res1 = await factory.app.delete(`/urls/${nonExistentId}`);
+      expect(res1.status).toBe(HTTP_ERRORS.NOT_FOUND);
+    });
+
+    it('should return an error if the url id is invalid', async () => {
+      const invalidId = '609c6a0f0ac8cd1e58fd6d87ddddd';
+      const res1 = await factory.app.delete(`/urls/${invalidId}`);
+      expect(res1.status).toBe(HTTP_ERRORS.BAD_REQUEST);
+    });
```
Comment (greatertomi) [inline src/test/url.test.ts]: Add more assertions to this error tests

**Review by PraiseTy — APPROVED** [review]

File: src/test/factory.ts (hunk 1 above)
Comment (PraiseTy) [reply src/test/factory.ts:38]: Done

File: src/test/url.test.ts (hunk 2 above)
Comment (PraiseTy) [reply src/test/url.test.ts:56]: Done. I didn't add the error messages to this

File: src/test/url.test.ts (hunk 4 above)
Comment (PraiseTy) [reply src/test/url.test.ts:151]: done

**Pull request merged**